src/emd/rt_bse.F

```diff
@@ -0,0 +1,1420 @@
+         !      rho(t+dt, 0) = rho_M
+         !      for j in 1,max_self_iter
+         !              rho(t+dt,j) = exp(- i S^(-1) H[rho(t+dt,j-1)] dt/2) rho_M exp(i H [rho(t+dt,j-1)] S^(-1) dt/2)
+         !              if ||rho(t+dt,j) - rho(t+dt,j-1)|| < epsilon
+         !                      break
+         ! *************** Determine rho_M ***************
+         ! Update the effective Hamiltonian
+         !  - sets the correct external field, updates Hartree and COHSEX terms to current density
+         CALL update_effective_ham(rtbse_env, rtbse_env%rho)
+         ! Convert the effective hamiltonian into the exponential
+         ! TODO : Store the result more explicitly - explicit dependence on input/output matrices
+         CALL ham_to_exp(rtbse_env)
+         ! Propagate the density to mid-point
+         CALL propagate_density(rtbse_env, rtbse_env%ham_workspace, rtbse_env%rho, rtbse_env%rho_M)
+         ! In initial iteration, copy rho_M to rho_new_last - that is our initial guess
+         DO j = 1, rtbse_env%n_spin
+            CALL cp_cfm_to_cfm(rtbse_env%rho_M(j)%matrix, rtbse_env%rho_new_last(j)%matrix)
+         END DO
+         ! *********** Start the self-consistent loop ************************
+         ! The guess Hamiltonian uses the guess density and field from the next timestep
+         rtbse_env%sim_time = REAL(i + 1, dp)*rtbse_env%sim_dt
+         rtbse_env%sim_step = i + 1
+         converged = .FALSE.
+         ! Prints the grid for etrs info dumping - only for log level > medium
+         CALL print_etrs_info_header(rtbse_env)
+         DO k = 1, rtbse_env%etrs_max_iter
+            CALL update_effective_ham(rtbse_env, rtbse_env%rho_new_last)
+            CALL ham_to_exp(rtbse_env)
+            CALL propagate_density(rtbse_env, rtbse_env%ham_workspace, rtbse_env%rho_M, rtbse_env%rho_new)
+            ! *** Self-consistency check ***
+            metric = rho_metric(rtbse_env%rho_new, rtbse_env%rho_new_last, rtbse_env%n_spin)
+            ! ETRS info - only for log level > medium
+            CALL print_etrs_info(rtbse_env, k, metric)
+            IF (metric < rtbse_env%etrs_threshold) THEN
+               converged = .TRUE.
+               EXIT
+            ELSE
+               ! Copy rho_new to rho_new_last
+               DO j = 1, rtbse_env%n_spin
+                  ! Leaving for free convergence
+                  CALL cp_cfm_to_cfm(rtbse_env%rho_new(j)%matrix, rtbse_env%rho_new_last(j)%matrix)
+               END DO
+            END IF
+         END DO
+         ! Not all of these are used, but they are all good metrics to check the convergence in problematic cases
+         CALL get_electron_number(rtbse_env, rtbse_env%rho_new, enum_re, enum_im)
+         ! TODO : Allow for conditional warning
+         CALL get_idempotence_deviation(rtbse_env, rtbse_env%rho_new, idempotence_dev)
+         ! DO j = 1, rtbse_env%n_spin
```

Comment:
If this is debug code, wrap it in `IF (.FALSE.) THEN <your code block> END IF` to enable compiler checks. Otherwise remove it.

Reply:
I have encapsulated the code in IF (.FALSE.) statements, as requested.